pkg/database/migrations.go

```diff
@@ -406,6 +409,63 @@ func (db *Database) getMigrations(ctx context.Context) *gormigrate.Gormigrate {
 				return nil
 			},
 		},
+		{
+			ID: "00017-AddAPIKeyPreviewAuthApp",
+			Migrate: func(tx *gorm.DB) error {
+				logger.Infof("db migrations: migrating authapp")
+				return tx.AutoMigrate(AuthorizedApp{}).Error
+			},
+			Rollback: func(tx *gorm.DB) error {
+				return nil
+			},
+		},
+		{
+			ID: "00018-IncreaseAPIKeySize",
+			Migrate: func(tx *gorm.DB) error {
+				logger.Infof("db migrations: increasing API key size")
+				sql := "ALTER TABLE authorized_apps ALTER COLUMN api_key TYPE varchar(512)"
+				return tx.Exec(sql).Error
+			},
+			Rollback: func(tx *gorm.DB) error {
+				sql := "ALTER TABLE authorized_apps ALTER COLUMN api_key TYPE varchar(100)"
+				return tx.Exec(sql).Error
+			},
+		},
+		{
+			ID: "00019-HMACAPIKeys",
+			Migrate: func(tx *gorm.DB) error {
+				logger.Infof("db migrations: HMACing existing api keys")
+
+				var apps []AuthorizedApp
+				if err := tx.Model(AuthorizedApp{}).Find(&apps).Error; err != nil {
+					return err
+				}
+
+				for _, app := range apps {
+					// If the key has a preview, it's v2
+					if app.APIKeyPreview != "" {
+						continue
+					}
+
+					apiKeyPreview := app.APIKey[:6]
+					newAPIKey, err := db.hmacAPIKey(app.APIKey)
```

Comment:
My worry w/ this is making sure the DB migrate tool has the right HMAC config present

Reply:
It calls `config.Open` which _requires_ the HMAC key to be present. If it's not it'll fail opening the db.

Should we set these in our build scripts? I don't really have an opinion on the best way to get them into the migration step.

Reply:
3 thoughts


1 - getting the secret ref into the build script would be OK
2 - We could deploy DB migration as a private cloud run service and CURL it w/ the right IAM permissions. Then it would be easy to terraform in the right data.
3 - Move this upgrade into the main server and handle it live on server startup. Can use the cleanup lock thing to run it once and set the next cleanup to occur years out. Then, say 2 releases out, we just remove this.